### PR TITLE
fix(report): eliminate N+1 query in patient report test description fetching

### LINKED ISSUE #3292: fix: N+1 query in PatientReportServiceImpl when resolving testDescription column
## Problem

In `PatientReportServiceImpl.mapToReportingData()`, the method `getCellValue()` 
is called for every result row × every column. When the `testDescription` column 
is present, it calls `testService.getTestById()` individually per row:
```java
case "testDescription":
    if (item.getTestId() != null) {
        Test test = testService.getTestById(item.getTestId()); // DB hit per row
        if (test != null && test.getDescription() != null) {
            return test.getDescription();
        }
    }
```

For a patient with N result rows, this results in N separate database queries 
just for test descriptions — a classic N+1 query problem.

Each `getTestById()` call opens its own Hibernate Session (the DAO is 
`@Transactional(readOnly=true)`), so the first-level cache provides no help 
across calls.

## Impact

- A patient with 50 test results triggers 50 extra DB round-trips
- Performance degrades linearly with result count
- Silent — no error is thrown, just slow response times

## Solution

Collect all unique `testId` values upfront before the row-building loop, 
batch-fetch them in a single query via `testService.getTestsByIds()`, store 
results in a `Map<String, Test>` for O(1) lookup, and pass the map into 
`getCellValue()`.
```java
// Before the loop — one DB call
Set<String> testIds = new HashSet<>();
for (TestResultItem item : results) {
    if (item.getTestId() != null) testIds.add(item.getTestId());
}
Map<String, Test> testsMap = new HashMap<>();
for (Test test : testService.getTestsByIds(testIds)) {
    testsMap.put(test.getId(), test);
}

// Inside getCellValue() — zero DB calls
case "testDescription":
    Test test = testsMap.get(item.getTestId());
    return (test != null && test.getDescription() != null) 
        ? test.getDescription() : "";
```

## Files Affected

-`src/main/java/org/openelisglobal/test/dao/TestDAO.java `
-`src/main/java/org/openelisglobal/test/daoimpl/TestDAOImpl.java `
-`src/main/java/org/openelisglobal/test/service/TestService.java `
-`src/main/java/org/openelisglobal/test/service/TestServiceImpl.java`

### PR DESCRIPTION
**Pull request by Junaidarmaan**
##  Problem

The patient report generation logic had an **N+1 query issue** while fetching test descriptions.

Inside the report mapping (`getCellValue`), the method `testService.getTestById()` was being called for each `TestResultItem`.  
This resulted in:

- 1 query to fetch results
- N additional queries to fetch test data (one per row)

### Impact
- Increased database load
- Redundant queries for repeated test IDs
- Performance degradation for large datasets

---

## ✅ Solution

This PR eliminates the N+1 query by introducing **batch fetching** for tests.

### Changes made:

- Added new method:
  - `getTestsByIds(Set<String> ids)` in:
    - `TestService`
    - `TestServiceImpl`
    - `TestDAO`
    - `TestDAOImpl`

- Fetch all required tests in a **single query** using `IN (:ids)`

- Converted fetched results into a `Map<String, Test>` for **O(1) lookup**

- Updated report generation logic:
  - Replaced per-item DB calls with map lookup
  - Modified `getCellValue` to accept `testsMap`

---

## 🚀 Result

| Before | After |
|-------|------|
| N + 1 queries ❌ | 2 queries ✅ |
| Repeated DB hits ❌ | Single batch fetch ✅ |
| O(N) DB access ❌ | O(1) lookup ✅ |

---

## 🧪 Notes

- Ensured null safety while collecting `testIds`
- Followed existing DAO + Service delegation pattern
- Used HQL (`IN :ids`) consistent with current codebase

---

## 🔗 Related Issue

Fixes #3292

### CODE REVIEW

File: src/main/java/org/openelisglobal/report/service/impl/PatientReportServiceImpl.java [hunk 1]
```diff
@@ -166,6 +171,22 @@ private ReportingData mapToReportingData(List<TestResultItem> results, Patient p
...
         List<ReportRow> rows = new ArrayList<>();
+
+        // creating and populating a set to store all testids
+        Set<String> testIds = new HashSet<>();
+        for (TestResultItem item : results) {
+            testIds.add(item.getTestId());
+        }
+
+        // fetching all tests by testids
+        List<Test> tests = testService.getTestsByIds(testIds);
```
[the remark this answers is not in the document]
Comment (Junaidarmaan) [reply src/main/java/org/openelisglobal/report/service/impl/PatientReportServiceImpl.java:185]: yeah i have fixed it thanks for pointing out

**Comment by Junaidarmaan** [comment]
@pmanko 
Hi, I noticed the E2E workflow is failing. From the logs, it looks like there might be a dependency resolution issue related to `org.opensaml:opensaml-saml-impl`.
Also, I’m not sure if any workflow approval is still required for fork PRs.
Could you please take a look or re-run the workflow when convenient? Happy to make any changes if needed.

File: src/main/java/org/openelisglobal/report/service/impl/PatientReportServiceImpl.java [hunk 2]
```diff
@@ -1,7 +1,12 @@
 package org.openelisglobal.report.service.impl;
```
Comment (Agaba-derrick) [inline src/main/java/org/openelisglobal/report/service/impl/PatientReportServiceImpl.java:2]: can we  update this branch  ?

**Review by Agaba-derrick — CHANGES_REQUESTED** [review]

**Review by mozzy11 — COMMENTED** [review]
Thanks @Junaidarmaan — this is a real N+1: on develop, getCellValue calls testService.getTestById per TestResultItem (PatientReportServiceImpl:220), and your batch getTestsByIds threaded properly through DAO → Service with a prefetched map is the right architectural fix. Before review, please rebase onto latest develop so the full backend CI runs (only the trigger job executed here), run mvn spotless:apply, and add a test covering getTestsByIds — including the empty-set input, which HQL IN-clauses handle badly if unguarded. Also double-check the report output is byte-identical for rows whose testId is missing from the map versus the old null-handling. With CI green and a test, this is a strong merge candidate.

**Comment by Agaba-derrick** [comment]
@Junaidarmaan are you able to adress the CI reviews here ?

**Review by mozzy11 — COMMENTED** [review]
Re-reviewed after the July 23-26 updates — all asks from my July 17 review are addressed:

- Confirmed the N+1 still exists on develop (`PatientReportServiceImpl` calls `testService.getTestById` per `TestResultItem` in `getCellValue`), so the fix is still needed.
- Null test IDs are now filtered before the batch fetch, the DAO guards the empty-set case before binding the HQL `IN` clause, and `TestServiceTest` covers both the matching-IDs and empty-set paths against the shared `testdata/test.xml` fixture (tests 1/2 = Blood/Urine Test — verified the fixture supports the assertions).
- Spotless applied, branch merged up through develop (Jul 26), and the earlier overlap with #3291 (`ReportServiceImpl`) is no longer in this diff.

Remaining process items, no code changes requested: @Agaba-derrick, your changes-requested review asked for a branch update, which is done — could you re-review? The full backend pipeline also hasn't run on this head yet (only the trigger job); we'll approve the workflow run so Build + Test executes the new `getTestsByIds` tests.
